File: Principles of Information Security/EV1/code.cpp
#include<bits/stdc++.h>
using namespace std;
typedef long long LL;
#define fi first
#define se second

LL po(LL a, LL n, LL p){
  if(n==0) return 1;
  LL an = po(a,n/2,p);
  an = (an*an)%p;
  if(n&1) return an*a%p;
  return an;
}

bool isprime(LL n) {
  int tc = 1000;
  if(n==1) return 0;
  if(tc>=n) tc = n-1;
  forn(i,tc){
    LL range = (n-1);
    LL num = rand() % range + 1;
    if(po(num, n-1, n)!=1)
      return 0;
  }
    return 1;
}
    

auto signature(int n) -> pair<LL,LL> {
   // LL k = rand();
   LL k = 2;
   // g is generator, p is prime...
   /*
    * generate x also. 
    */
   LL r = po(g,k,p);
   LL e = hashing(conc(r,m));
   LL s = k - x*e;
   return {s,e};
}

// rv = g.s y.e
auto verify(pair<int,int> sign) ->bool{
  LL s = sign.fi;
  LL e = sign.se;
  LL rv = (po(g,s,p)*po(y,e,p))%p;
  LL ev = hashing(conc(rv,m));
  if(ev == e){
    cout<<"Verified"<<endl;
  }
  else cout<<"Not Verified"<<endl;
}

auto generate(int n) -> {
    
}

int main(){
  int n; cin>>n;
  
}

File: Principles of Information Security/EV1/code.py
from random import randint
import sympy as sp

numbits = 10

def po(a,n,p):
    an = 1
    a = a%p
    if a==0:
        return 0
    while(n):
        if(n&1):
            an = an*a%p
        n = n>>1
        an = an*an%p
    return an

def sign(g, p, x):
    rn = randint(2, p-1)
    t = po(g, rn, p)
    y = po(g, x, p)
    c = hashfun(g, p, y)
    z = c*x + rn
    return t, y, c, z

def verify(g, p, z, t, y):
    c = hashfun(g,p,y)
    if po(g,z,p) == ((t*po(y,c,p))%p):
        return True
    return False

def generator(p , f):
    phi = p - 1
    for x in range(2, p+1):
        tr = True
        for fac in [f, 2]:
            if po(x, phi/fac, p)==1:
                tr = False
                break
        if tr:
            return x
    return -1

def generate_p(n):
    while True:
        pr = randint(2**(n-1), 2**n)
        if sp.isprime(pr) and sp.isprime(2*pr+1):
            return 2*pr+1, pr

# input nbits
nbits = int(input("Enter number of bits: "))
print("Number of bits: ",numbits) 
prime, fac = generate_p(numbits)
print("Prime : ", prime)
gen = generator(prime, fac)



